model/src/lib.rs

```diff
@@ -52,26 +58,97 @@ pub struct OrderCreation {
 }
 
 impl OrderCreation {
+    pub const ORDER_TYPE_HASH: [u8; 32] =
+        hex!("b71968fcf5e55b9c3370f2809d4078a4695be79dfa43e5aa1f2baa0a9b84f186");
+
     pub fn token_pair(&self) -> Option<TokenPair> {
         TokenPair::new(self.buy_token, self.sell_token)
     }
-    pub fn order_owner(&self) -> H160 {
-        // dummy implementation, waiting for Valentins PR
-        H160::zero()
+
+    pub fn order_digest(&self) -> [u8; 32] {
+        let mut hash_data = [0u8; 320];
+        hash_data[0..32].copy_from_slice(&Self::ORDER_TYPE_HASH);
+        // Some slots are not assigned (stay 0) because all values are extended to 256 bits.
+        hash_data[44..64].copy_from_slice(self.sell_token.as_fixed_bytes());
+        hash_data[76..96].copy_from_slice(self.buy_token.as_fixed_bytes());
+        self.sell_amount.to_big_endian(&mut hash_data[96..128]);
+        self.buy_amount.to_big_endian(&mut hash_data[128..160]);
+        hash_data[188..192].copy_from_slice(&self.valid_to.to_be_bytes());
+        hash_data[220..224].copy_from_slice(&self.app_data.to_be_bytes());
+        self.fee_amount.to_big_endian(&mut hash_data[224..256]);
+        hash_data[287] = match self.order_kind {
+            OrderKind::Sell => 0,
+            OrderKind::Buy => 1,
+        };
+        hash_data[319] = self.partially_fillable as u8;
+        signing::keccak256(&hash_data)
+    }
+
+    pub fn signing_digest_typed_data(&self, domain_separator: &[u8; 32]) -> [u8; 32] {
+        let mut hash_data = [0u8; 66];
+        hash_data[0..2].copy_from_slice(&[0x19, 0x01]);
+        hash_data[2..34].copy_from_slice(domain_separator);
+        hash_data[34..66].copy_from_slice(&self.order_digest());
+        signing::keccak256(&hash_data)
+    }
+
+    pub fn signing_digest_message(&self, domain_separator: &[u8; 32]) -> [u8; 32] {
+        let mut hash_data = [0u8; 92];
+        hash_data[0..28].copy_from_slice(b"\x19Ethereum Signed Message:\n64");
+        hash_data[28..60].copy_from_slice(domain_separator);
+        hash_data[60..92].copy_from_slice(&self.order_digest());
+        signing::keccak256(&hash_data)
+    }
+
+    pub fn signing_digest(&self, domain_separator: &[u8; 32]) -> [u8; 32] {
+        if self.signature.v & 0x80 == 0 {
+            self.signing_digest_typed_data(domain_separator)
+        } else {
+            self.signing_digest_message(domain_separator)
+        }
     }
-    pub fn order_digest(&self) -> H256 {
-        // dummy implementation, waiting for Valentins PR
-        H256::zero()
+
+    // If signature is valid returns the owner.
+    pub fn validate_signature(&self, domain_separator: &[u8; 32]) -> Option<H160> {
+        let v = self.signature.v & 0x1f;
```

Comment:
why this? Could you please add a comment?

Reply:
Its because the signing scheme is encoded in the signature. Maybe this can be turned into an abstraction so that its clearer? Something like:
```
enum SigningScheme {
    Eip712TypedData,
    Message,
}
```

Reply:
With a comment linking GPv@Encoding.sol I don't think we have to explain this in more detail here. This code doesn't make any design decisions. It just copies the implementation in the smart contract.

Reply:
Personally I would prefer if the implementation referenced the EIP-712 which is the signing standard used instead of the SC (which just adds the detail of also supporting wallets that don't support signing typed data) maybe using the JS code from the same repo as reference (as that is what the solidity code is tested against). My main argument being that this would make the Rust code more expressive and self-documenting than if it just ported the Solidity implementation, which is a bit low level to get around Solidity/EVM limitations and not necessarily because it is the cleanest implementation possible. In fact, it could be implemented as `keccack256(abi.encode(...))` if we weren't going for gas optimizations.

Reply:
I don't want to block merging this PR. I imagine this is also in some ways related to refactoring and #96 and #97 so we can merge as is IMO.